File: go/types/path_test.go
// Copyright 2016 The Noms Authors. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

package types

import (
	"fmt"
	"testing"

	"github.com/attic-labs/noms/go/hash"
	"github.com/attic-labs/testify/assert"
)

func assertPathResolvesTo(assert *assert.Assertions, expect, ref Value, p Path) {
	actual := p.Resolve(ref)
	if expect == nil {
		assert.Nil(actual)
	} else if actual == nil {
		assert.Fail("", "Expected %s, but got nil", EncodedValue(expect))
	} else {
		assert.True(expect.Equals(actual), "Expected %s, but got %s", EncodedValue(expect), EncodedValue(actual))
	}
}

func assertPathStringResolvesTo(assert *assert.Assertions, expect, ref Value, str string) {
	p, err := NewPath().AddPath(str)
	assert.NoError(err)
	assertPathResolvesTo(assert, expect, ref, p)
}

func TestPathStruct(t *testing.T) {
	assert := assert.New(t)

	v := NewStruct("", structData{
		"foo": String("foo"),
		"bar": Bool(false),
		"baz": Number(203),
	})

	assertPathResolvesTo(assert, String("foo"), v, NewPath().AddField("foo"))
	assertPathStringResolvesTo(assert, String("foo"), v, `.foo`)
	assertPathResolvesTo(assert, Bool(false), v, NewPath().AddField("bar"))
	assertPathStringResolvesTo(assert, Bool(false), v, `.bar`)
	assertPathResolvesTo(assert, Number(203), v, NewPath().AddField("baz"))
	assertPathStringResolvesTo(assert, Number(203), v, `.baz`)
	assertPathResolvesTo(assert, nil, v, NewPath().AddField("notHere"))
	assertPathStringResolvesTo(assert, nil, v, `.notHere`)
}

func TestPathList(t *testing.T) {
	assert := assert.New(t)

	v := NewList(Number(1), Number(3), String("foo"), Bool(false))

	assertPathResolvesTo(assert, Number(1), v, NewPath().AddIndex(Number(0)))
	assertPathStringResolvesTo(assert, Number(1), v, `[0]`)
	assertPathResolvesTo(assert, Number(3), v, NewPath().AddIndex(Number(1)))
	assertPathStringResolvesTo(assert, Number(3), v, `[1]`)
	assertPathResolvesTo(assert, String("foo"), v, NewPath().AddIndex(Number(2)))
	assertPathStringResolvesTo(assert, String("foo"), v, `[2]`)
	assertPathResolvesTo(assert, Bool(false), v, NewPath().AddIndex(Number(3)))
	assertPathStringResolvesTo(assert, Bool(false), v, `[3]`)
	assertPathResolvesTo(assert, nil, v, NewPath().AddIndex(Number(4)))
	assertPathStringResolvesTo(assert, nil, v, `[4]`)
	assertPathResolvesTo(assert, nil, v, NewPath().AddIndex(Number(-4)))
	assertPathStringResolvesTo(assert, nil, v, `[-4]`)
}

func TestPathMap(t *testing.T) {
	assert := assert.New(t)

	v := NewMap(
		Number(1), String("foo"),
		String("two"), String("bar"),
		Bool(false), Number(23),
		Number(2.3), Number(4.5),
	)

	assertPathResolvesTo(assert, String("foo"), v, NewPath().AddIndex(Number(1)))
	assertPathStringResolvesTo(assert, String("foo"), v, `[1]`)
	assertPathResolvesTo(assert, String("bar"), v, NewPath().AddIndex(String("two")))
	assertPathStringResolvesTo(assert, String("bar"), v, `["two"]`)
	assertPathResolvesTo(assert, Number(23), v, NewPath().AddIndex(Bool(false)))
	assertPathStringResolvesTo(assert, Number(23), v, `[false]`)
	assertPathResolvesTo(assert, Number(4.5), v, NewPath().AddIndex(Number(2.3)))
	assertPathStringResolvesTo(assert, Number(4.5), v, `[2.3]`)
	assertPathResolvesTo(assert, nil, v, NewPath().AddIndex(Number(4)))
	assertPathStringResolvesTo(assert, nil, v, `[4]`)
}

func TestPathHashIndex(t *testing.T) {
	assert := assert.New(t)

	b := Bool(true)
	br := NewRef(b)
	i := Number(0)
	str := String("foo")
	l := NewList(b, i, str)
	lr := NewRef(l)
	m := NewMap(
		b, br,
		br, i,
		i, str,
		l, lr,
		lr, b,
	)
	s := NewSet(b, br, i, str, l, lr)

	hashStr := func(v Value) string {
		return fmt.Sprintf("[#%s]", v.Hash())
	}

	resolvesTo := func(col, exp, val Value) {
		assertPathResolvesTo(assert, exp, col, NewPath().AddHashIndex(val.Hash()))
		assertPathStringResolvesTo(assert, exp, col, hashStr(val))
	}

	// Primitives are only addressable by their values.
	resolvesTo(m, nil, b)
	resolvesTo(m, nil, i)
	resolvesTo(m, nil, str)
	resolvesTo(s, nil, b)
	resolvesTo(s, nil, i)
	resolvesTo(s, nil, str)

	// Other values are only addressable by their hashes.

	resolvesTo(m, i, br)
	resolvesTo(m, lr, l)
	resolvesTo(m, b, lr)
	resolvesTo(s, br, br)
	resolvesTo(s, l, l)
	resolvesTo(s, lr, lr)

	// Lists cannot be addressed by hashes, obviously.
	resolvesTo(l, nil, i)
}

func TestPathHashIndexOfSingletonCollection(t *testing.T) {
	// This test is to make sure we don't accidentally return |b| if it's the only element.
	assert := assert.New(t)

	resolvesToNil := func(col, val Value) {
		assertPathResolvesTo(assert, nil, col, NewPath().AddHashIndex(val.Hash()))
		assertPathStringResolvesTo(assert, nil, col, fmt.Sprintf("[#%s]", val.Hash()))
	}

	b := Bool(true)
	resolvesToNil(NewMap(b, b), b)
	resolvesToNil(NewSet(b), b)
}

func TestPathMulti(t *testing.T) {
	assert := assert.New(t)

	m1 := NewMap(
		String("a"), String("foo"),
		String("b"), String("bar"),
		String("c"), String("car"),
	)

	m2 := NewMap(
		String("d"), String("dar"),
		Bool(false), String("earth"),
		m1, String("fire"),
	)

	l := NewList(m1, m2)

	s := NewStruct("", structData{
		"foo": l,
	})

	assertPathResolvesTo(assert, l, s, NewPath().AddField("foo"))
	assertPathStringResolvesTo(assert, l, s, `.foo`)
	assertPathResolvesTo(assert, m1, s, NewPath().AddField("foo").AddIndex(Number(0)))
	assertPathStringResolvesTo(assert, m1, s, `.foo[0]`)
	assertPathResolvesTo(assert, String("foo"), s, NewPath().AddField("foo").AddIndex(Number(0)).AddIndex(String("a")))
	assertPathStringResolvesTo(assert, String("foo"), s, `.foo[0]["a"]`)
	assertPathResolvesTo(assert, String("bar"), s, NewPath().AddField("foo").AddIndex(Number(0)).AddIndex(String("b")))
	assertPathStringResolvesTo(assert, String("bar"), s, `.foo[0]["b"]`)
	assertPathResolvesTo(assert, String("car"), s, NewPath().AddField("foo").AddIndex(Number(0)).AddIndex(String("c")))
	assertPathStringResolvesTo(assert, String("car"), s, `.foo[0]["c"]`)
	assertPathResolvesTo(assert, nil, s, NewPath().AddField("foo").AddIndex(Number(0)).AddIndex(String("x")))
	assertPathStringResolvesTo(assert, nil, s, `.foo[0]["x"]`)
	assertPathResolvesTo(assert, nil, s, NewPath().AddField("foo").AddIndex(Number(2)).AddIndex(String("c")))
	assertPathStringResolvesTo(assert, nil, s, `.foo[2]["c"]`)
	assertPathResolvesTo(assert, nil, s, NewPath().AddField("notHere").AddIndex(Number(0)).AddIndex(String("c")))
	assertPathStringResolvesTo(assert, nil, s, `.notHere[0]["c"]`)
	assertPathResolvesTo(assert, m2, s, NewPath().AddField("foo").AddIndex(Number(1)))
	assertPathStringResolvesTo(assert, m2, s, `.foo[1]`)
	assertPathResolvesTo(assert, String("dar"), s, NewPath().AddField("foo").AddIndex(Number(1)).AddIndex(String("d")))
	assertPathStringResolvesTo(assert, String("dar"), s, `.foo[1]["d"]`)
	assertPathResolvesTo(assert, String("earth"), s, NewPath().AddField("foo").AddIndex(Number(1)).AddIndex(Bool(false)))
	assertPathStringResolvesTo(assert, String("earth"), s, `.foo[1][false]`)
	assertPathResolvesTo(assert, String("fire"), s, NewPath().AddField("foo").AddIndex(Number(1)).AddHashIndex(m1.Hash()))
	assertPathStringResolvesTo(assert, String("fire"), s, fmt.Sprintf(`.foo[1][#%s]`, m1.Hash().String()))
}

func TestPathToAndFromString(t *testing.T) {
	assert := assert.New(t)

	test := func(str string, p Path) {
		assert.Equal(str, p.String())
		p2, err := NewPath().AddPath(str)
		assert.NoError(err)
		assert.Equal(p, p2)
	}

	test("[0]", NewPath().AddIndex(Number(0)))
	test("[\"0\"][\"1\"][\"100\"]", NewPath().AddIndex(String("0")).AddIndex(String("1")).AddIndex(String("100")))
	test(".foo[0].bar[4.5][false]", NewPath().AddField("foo").AddIndex(Number(0)).AddField("bar").AddIndex(Number(4.5)).AddIndex(Bool(false)))
	h := Number(42).Hash() // arbitrary hash
	test(fmt.Sprintf(".foo[#%s]", h.String()), NewPath().AddField("foo").AddHashIndex(h))
	test(fmt.Sprintf("foo.bar[#%s]", h.String()), NewPath().SetRootDataset("foo").AddField("bar").AddHashIndex(h))
	test(fmt.Sprintf("#%s.bar[42]", h.String()), NewPath().SetRootHash(h).AddField("bar").AddIndex(Number(42)))
}

func TestPathImmutability(t *testing.T) {
	assert := assert.New(t)
	p1 := NewPath().AddField("/").AddField("value").AddField("data").AddIndex(Number(1)).AddField("data")
	p2 := p1.AddField("x")
	p3 := p1.AddField("y")
	p4 := p3.AddIndex(Number(19))
	assert.Equal("./.value.data[1].data", p1.String())
	assert.Equal("./.value.data[1].data.x", p2.String())
	assert.Equal("./.value.data[1].data.y", p3.String())
	assert.Equal("./.value.data[1].data.y[19]", p4.String())
}

func TestPathParseSuccess(t *testing.T) {
	assert := assert.New(t)

	test := func(str string, expectPath Path) {
		p, err := NewPath().AddPath(str)
		assert.NoError(err)
		assert.Equal(expectPath, p)
	}

	test(".foo", NewPath().AddField("foo"))
	test(".Q", NewPath().AddField("Q"))
	test(".QQ", NewPath().AddField("QQ"))
	test("[true]", NewPath().AddIndex(Bool(true)))
	test("[false]", NewPath().AddIndex(Bool(false)))
	test("[42]", NewPath().AddIndex(Number(42)))
	test("[1e4]", NewPath().AddIndex(Number(1e4)))
	test("[1.]", NewPath().AddIndex(Number(1.)))
	test("[1.345]", NewPath().AddIndex(Number(1.345)))
	test(`[""]`, NewPath().AddIndex(String("")))
	test(`["42"]`, NewPath().AddIndex(String("42")))
	test("[\"line\nbreak\rreturn\"]", NewPath().AddIndex(String("line\nbreak\rreturn")))
	test(`["qu\\ote\""]`, NewPath().AddIndex(String(`qu\ote"`)))
	test(`["π"]`, NewPath().AddIndex(String("π")))
	test(`["[[br][]acke]]ts"]`, NewPath().AddIndex(String("[[br][]acke]]ts")))
	test(`["xπy✌z"]`, NewPath().AddIndex(String("xπy✌z")))
	test(`["ಠ_ಠ"]`, NewPath().AddIndex(String("ಠ_ಠ")))
	test(`["ಠ_ಠ"]`, NewPath().AddIndex(String("ಠ_ಠ")))
}

func TestPathParseErrors(t *testing.T) {
	assert := assert.New(t)

	test := func(str, expectError string) {
		p, err := NewPath().AddPath(str)
		assert.Equal(Path{}, p)
		if err != nil {
			assert.Equal(expectError, err.Error())
		} else {
			assert.Fail("Expected " + expectError)
		}
	}

	test("", "Empty path")
	test(".", "Invalid field: ")
	test("[", "Path ends in [")
	test(".#", "Invalid field: #")
	test(". ", "Invalid field:  ")
	test(". invalid.field", "Invalid field:  invalid.field")
	test(".foo.", "Invalid field: ")
	test(".foo.#invalid.field", "Invalid field: #invalid.field")
	test(".foo!", "Invalid operator: !")
	test(".foo!bar", "Invalid operator: !")
	test(".foo#", "# operator can only be the first character")
	test(".foo#bar", "# operator can only be the first character")
	test(".foo[", "Path ends in [")
	test(".foo[.bar", "[ is missing closing ]")
	test(".foo]", "] is missing opening [")
	test(".foo].bar", "] is missing opening [")
	test(".foo[]", "Empty index value")
	test(".foo[[]", "Invalid index: [")
	test(".foo[[]]", "Invalid index: [")
	test(".foo[42.1.2]", "Invalid index: 42.1.2")
	test(".foo[1f4]", "Invalid index: 1f4")
	test(".foo[hello]", "Invalid index: hello")
	test(".foo['hello']", "Invalid index: 'hello'")
	test(`.foo[\]`, `Invalid index: \`)
	test(`.foo[\\]`, `Invalid index: \\`)
	test(`.foo["hello]`, "[ is missing closing ]")
	test(`.foo["hello`, "[ is missing closing ]")
	test(`.foo["`, "[ is missing closing ]")
	test(`.foo["\`, "[ is missing closing ]")
	test(`.foo["]`, "[ is missing closing ]")
	test(".foo[#]", "Invalid hash: ")
	test(".foo[#sha1-invalid]", "Invalid hash: sha1-invalid")
	test(`.foo["hello\nworld"]`, `Only " and \ can be escaped`)
	test(".foo[42]bar", "Invalid operator: b")
	test("#foo", "Invalid hash: foo")
	test("!foo", "Invalid dataset name: !foo")
}

type testPathRootGetter struct {
	datasets map[string]Value
	hashes   map[hash.Hash]Value
}

func (g testPathRootGetter) GetDatasetHead(id string) Value {
	return g.datasets[id]
}

func (g testPathRootGetter) GetHash(h hash.Hash) Value {
	return g.hashes[h]
}

func TestFullPaths(t *testing.T) {
	assert := assert.New(t)

	s0, s1 := String("foo"), String("bar")
	list := NewList(s0, s1)
	emptySet := NewSet()
	ds := NewStruct("Commit", structData{
		"parents": emptySet,
		"value":   list,
	})

	g := testPathRootGetter{
		map[string]Value{"ds": ds},
		map[hash.Hash]Value{s0.Hash(): s0, s1.Hash(): s1, list.Hash(): list},
	}

	resolvesTo := func(exp Value, str string) {
		p, err := ParsePath(str)
		assert.NoError(err)
		act := p.ResolveFromRoot(g)
		if exp == nil {
			assert.Nil(act)
		} else {
			assert.True(exp.Equals(act))
		}
	}

	resolvesTo(ds, "ds")
	resolvesTo(emptySet, "ds.parents")
	resolvesTo(list, "ds.value")
	resolvesTo(s0, "ds.value[0]")
	resolvesTo(s1, "ds.value[1]")
	resolvesTo(list, "#"+list.Hash().String())
	resolvesTo(s0, "#"+s0.Hash().String())
	resolvesTo(s1, "#"+s1.Hash().String())
	resolvesTo(s0, "#"+list.Hash().String()+"[0]")
	resolvesTo(s1, "#"+list.Hash().String()+"[1]")

	resolvesTo(nil, "foo")
	resolvesTo(nil, "foo.parents")
	resolvesTo(nil, "foo.value")
	resolvesTo(nil, "foo.value[0]")
	resolvesTo(nil, "#"+String("baz").Hash().String())
	resolvesTo(nil, "#"+String("baz").Hash().String()+"[0]")
}

File: js/src/path-test.js
// @flow

// Copyright 2016 Attic Labs, Inc. All rights reserved.
// Licensed under the Apache License, version 2.0:
// http://www.apache.org/licenses/LICENSE-2.0

import {assert} from 'chai';
import {suite, test} from 'mocha';
import {equals} from './compare.js';

import {getHash} from './get-hash.js';
import List from './list.js';
import Map from './map.js';
import {default as Path} from './path.js';
import Ref from './ref.js';
import Set from './set.js';
import type Value from './value.js';
import {newStruct} from './struct.js';

function strify(s) {
  return JSON.stringify(s);
}

function hashIdx(v: Value): string {
  return `[#${getHash(v).toString()}]`;
}

async function assertResolvesTo(expect: ?Value, ref: Value, str: string) {
  const p = Path.parse(str);
  const actual = await p.resolve(ref);
  if (expect == null) {
    assert.isTrue(actual == null, `Expected null, but got ${strify(actual)}`);
  } else if (actual == null) {
    assert.isTrue(false, `Expected ${strify(expect)}, but got null`);
  } else {
    assert.isTrue(equals(expect, actual), `Expected ${strify(expect)}, but got ${strify(actual)}`);
  }
}

suite('Path', () => {
  test('struct', async () => {
    const v = newStruct('', {
      foo: 'foo',
      bar: false,
      baz: 203,
    });

    await assertResolvesTo('foo', v, '.foo');
    await assertResolvesTo(false, v, '.bar');
    await assertResolvesTo(203, v, '.baz');
    await assertResolvesTo(null, v, '.notHere');

    const v2 = newStruct('', {
      v1: v,
    });

    await assertResolvesTo('foo', v2, '.v1.foo');
    await assertResolvesTo(false, v2, '.v1.bar');
    await assertResolvesTo(203, v2, '.v1.baz');
    await assertResolvesTo(undefined, v2, '.v1.notHere');
    await assertResolvesTo(undefined, v2, '.notHere.foo');
  });

  test('index', async () => {
    let v: Value;
    const resolvesTo = async (exp: ?Value, val: Value, str: string) => {
      // Indices resolve to |exp|.
      await assertResolvesTo(exp, v, str);
      // Keys resolves to themselves.
      if (exp != null) {
        exp = val;
      }
      await assertResolvesTo(exp, v, str + '@key');
    };

    v = new List([1, 3, 'foo', false]);

    await resolvesTo(1, 0, '[0]');
    await resolvesTo(3, 1, '[1]');
    await resolvesTo('foo', 2, '[2]');
    await resolvesTo(false, 3, '[3]');
    await resolvesTo(null, 4, '[4]');
    await resolvesTo(null, -4, '[-4]');

    v = new Map([
      [1, 'foo'],
      ['two', 'bar'],
      [false, 23],
      [2.3, 4.5],
    ]);

    await resolvesTo('foo', 1, '[1]');
    await resolvesTo('bar', 'two', '["two"]');
    await resolvesTo(23, false, '[false]');
    await resolvesTo(4.5, 2.3, '[2.3]');
    await resolvesTo(null, 4, '[4]');
  });

  test('hash index', async () => {
    const b = true;
    const br = new Ref(b);
    const i = 0;
    const str = 'foo';
    const l = new List([b, i, str]);
    const lr = new Ref(l);
    const m = new Map([
      [b, br],
      [br, i],
      [i, str],
      [l, lr],
      [lr, b],
    ]);
    const s = new Set([b, br, i, str, l, lr]);

    const resolvesTo = async (col: Value, exp: ?Value, val: Value) => {
      // Values resolve to |exp|.
      await assertResolvesTo(exp, col, hashIdx(val));
      // Keys resolves to themselves.
      if (exp != null) {
        exp = val;
      }
      await assertResolvesTo(exp, col, hashIdx(val) + '@key');
    };

    // Primitives are only addressable by their values.
    await resolvesTo(m, null, b);
    await resolvesTo(m, null, i);
    await resolvesTo(m, null, str);
    await resolvesTo(s, null, b);
    await resolvesTo(s, null, i);
    await resolvesTo(s, null, str);

    // Other values are only addressable by their hashes.
    await resolvesTo(m, i, br);
    await resolvesTo(m, lr, l);
    await resolvesTo(m, b, lr);
    await resolvesTo(s, br, br);
    await resolvesTo(s, l, l);
    await resolvesTo(s, lr, lr);

    // Lists cannot be addressed by hashes, obviously.
    await resolvesTo(l, null, i);
  });

  test('hash index of singleton collection', async () => {
    // This test is to make sure we don't accidentally return the element of a singleton map.
    const resolvesToNull = async (col: Value, v: Value) => {
      await assertResolvesTo(null, col, hashIdx(v));
    };

    await resolvesToNull(new Map([[true, true]]), true);
    await resolvesToNull(new Set([true]), true);
  });

  test('multi', async () => {
    const m1 = new Map([
      ['a', 'foo'],
      ['b', 'bar'],
      ['c', 'car'],
    ]);

    const m2 = new Map([
      ['d', 'dar'],
      [false, 'earth'],
      [m1, 'fire'],
    ]);

    const l = new List([m1, m2]);

    const s = newStruct('', {
      'foo': l,
    });

    await assertResolvesTo(l, s, '.foo');
    await assertResolvesTo(m1, s, '.foo[0]');
    await assertResolvesTo('foo', s, '.foo[0]["a"]');
    await assertResolvesTo('bar', s, '.foo[0]["b"]');
    await assertResolvesTo('car', s, '.foo[0]["c"]');
    await assertResolvesTo(null, s, '.foo[0]["x"]');
    await assertResolvesTo(null, s, '.foo[2]["c"]');
    await assertResolvesTo(null, s, '.notHere[0]["c"]');
    await assertResolvesTo(m2, s, '.foo[1]');
    await assertResolvesTo('dar', s, '.foo[1]["d"]');
    await assertResolvesTo('earth', s, '.foo[1][false]');
    await assertResolvesTo('fire', s, `.foo[1]${hashIdx(m1)}`);
    await assertResolvesTo(m1, s, `.foo[1]${hashIdx(m1)}@key`);
    await assertResolvesTo('car', s, `.foo[1]${hashIdx(m1)}@key["c"]`);
  });

  test('parse success', () => {
    const test = (s: string) => {
      const p = Path.parse(s);
      let expect = s;
      // Human readable serialization special cases.
      if (expect === '[1e4]') {
        expect = '[10000]';
      } else if (expect === '[1.]') {
        expect = '[1]';
      } else if (expect === '["line\nbreak\rreturn"]') {
        expect = '["line\\nbreak\\rreturn"]';
      }
      assert.strictEqual(expect, p.toString());
    };

    const h = getHash(42); // arbitrary hash

    test('.foo');
    test('.Q');
    test('.QQ');
    test('[true]');
    test('[false]');
    test('[false]@key');
    test('[42]');
    test('[42]@key');
    test('[1e4]');
    test('[1.]');
    test('[1.345]');
    test('[""]');
    test('["42"]');
    test('["42"]@key');
    test('[\"line\nbreak\rreturn\"]');
    test('["qu\\\\ote\\\""]');
    test('["π"]');
    test('["[[br][]acke]]ts"]');
    test('["xπy✌z"]');
    test('["ಠ_ಠ"]');
    test('["0"]["1"]["100"]');
    test('.foo[0].bar[4.5][false]');
    test(`.foo[#${h.toString()}]`);
    test(`.bar[#${h.toString()}]@key`);
  });

  test('parse errors', () => {
    const test = (s: string, expectErr: string) => {
      try {
        Path.parse(s);
        assert.isOk(false, 'Expected error: ' + expectErr);
      } catch (e) {
        assert.strictEqual(expectErr, e.message);
      }
    };

    test('', 'Empty path');
    test('.', 'Invalid field: ');
    test('[', 'Path ends in [');
    test(']', '] is missing opening [');
    test('.#', 'Invalid field: #');
    test('. ', 'Invalid field:  ');
    test('. invalid.field', 'Invalid field:  invalid.field');
    test('.foo.', 'Invalid field: ');
    test('.foo.#invalid.field', 'Invalid field: #invalid.field');
    test('.foo!', 'Invalid operator: !');
    test('.foo!bar', 'Invalid operator: !');
    test('.foo#', 'Invalid operator: #');
    test('.foo#bar', 'Invalid operator: #');
    test('.foo[', 'Path ends in [');
    test('.foo[.bar', '[ is missing closing ]');
    test('.foo]', '] is missing opening [');
    test('.foo].bar', '] is missing opening [');
    test('.foo[]', 'Empty index value');
    test('.foo[[]', 'Invalid index: [');
    test('.foo[[]]', 'Invalid index: [');
    test('.foo[42.1.2]', 'Invalid index: 42.1.2');
    test('.foo[1f4]', 'Invalid index: 1f4');
    test('.foo[hello]', 'Invalid index: hello');
    test('.foo[\'hello\']', 'Invalid index: \'hello\'');
    test('.foo[\\]', 'Invalid index: \\');
    test('.foo[\\\\]', 'Invalid index: \\\\');
    test('.foo["hello]', '[ is missing closing ]');
    test('.foo["hello', '[ is missing closing ]');
    test('.foo["', '[ is missing closing ]');
    test('.foo["\\', '[ is missing closing ]');
    test('.foo["]', '[ is missing closing ]');
    test('.foo[#]', 'Invalid hash: ');
    test('.foo[#invalid]', 'Invalid hash: invalid');
    test('.foo["hello\\nworld"]', 'Only " and \\ can be escaped');
    test('.foo[42]bar', 'Invalid operator: b');
    test('#foo', 'Invalid operator: #');
    test('!foo', 'Invalid operator: !');
    test('@foo', 'Invalid operator: @');
    test('@key', 'Invalid operator: @');
    test('.foo[42]@soup', 'Unsupported annotation: @soup');
  });
});
